src/types.jl

```diff
@@ -97,13 +52,416 @@ for Model in [:NeuralNetworkClassifier, :ImageClassifier]
+
+# Examples
+
+In this example we use MLJ to classify the MNIST image dataset
+```julia
+using MLJ
+using Flux
+import MLJFlux
+import MLJIteration # for `skip`
+
+MLJ.color_off()
+
+using Plots
+pyplot(size=(600, 300*(sqrt(5)-1)));
+```
+First we want to download the MNIST dataset, and unpack into images and labels
+```julia
+import MLDatasets: MNIST
+
+ENV["DATADEPS_ALWAYS_ACCEPT"] = true
+images, labels = MNIST.traindata();
+```
+In MLJ, integers cannot be used for encoding categorical data, so we must coerce them into the `Multiclass` [scientific type](https://juliaai.github.io/ScientificTypes.jl/dev/). For more in this, see [Working with Categorical Data](https://alan-turing-institute.github.io/MLJ.jl/dev/working_with_categorical_data/):
+```julia
+labels = coerce(labels, Multiclass);
+images = coerce(images, GrayImage);
+
+# Checking scientific types:
+
+@assert scitype(images) <: AbstractVector{<:Image}
+@assert scitype(labels) <: AbstractVector{<:Finite}
+
+images[1]
+```
+For general instructions on coercing image data, see [type coercion for image data](https://alan-turing-institute.github.io/ScientificTypes.jl/dev/%23Type-coercion-for-image-data-1)
+We start by defining a suitable `builder` object. This is a recipe
+for building the neural network. Our builder will work for images of
+any (constant) size, whether they be color or black and white (ie,
+single or multi-channel).  The architecture always consists of six
+alternating convolution and max-pool layers, and a final dense
+layer; the filter size and the number of channels after each
+convolution layer is customisable.
+```julia
+import MLJFlux
+
+struct MyConvBuilder
+    filter_size::Int
+    channels1::Int
+    channels2::Int
+    channels3::Int
+end
+
+make2d(x::AbstractArray) = reshape(x, :, size(x)[end])
+
+function MLJFlux.build(b::MyConvBuilder, rng, n_in, n_out, n_channels)
+    k, c1, c2, c3 = b.filter_size, b.channels1, b.channels2, b.channels3
+    mod(k, 2) == 1 || error("`filter_size` must be odd. ")
+    p = div(k - 1, 2) # padding to preserve image size
+    init = Flux.glorot_uniform(rng)
+    front = Chain(
+        Conv((k, k), n_channels => c1, pad=(p, p), relu, init=init),
+        MaxPool((2, 2)),
+        Conv((k, k), c1 => c2, pad=(p, p), relu, init=init),
+        MaxPool((2, 2)),
+        Conv((k, k), c2 => c3, pad=(p, p), relu, init=init),
+        MaxPool((2 ,2)),
+        make2d)
+    d = Flux.outputsize(front, (n_in..., n_channels, 1)) |> first
+    return Chain(front, Dense(d, n_out, init=init))
+end
+```
+It is important to note that in our `build` function, there is no final softmax. This is applie by default in all MLJFlux classifiers, using the `finaliser` hyperparameter of the classifier. Now that we have our builder defined, we can define the actual moel. If you have a GPU, you can substitute in `acceleration=CudaLibs()` below. Note that in the case of convolutions, this will **greatly** increase the speed of training.
```

Comment:
```suggestion
It is important to note that in our `build` function, there is no final `softmax`. This is applied by default in all MLJFlux classifiers (override this using the `finaliser` hyperparameter). Now that we have our builder defined, we can define the actual model. If you have a GPU, you can substitute in `acceleration=CudaLibs()` below. Note that in the case of convolutions, this will **greatly** increase the speed of training.
```